File: Satori-iOS/Satori/Services/AudioService.swift
import SwiftUI
import AVFoundation

// MARK: - 音频服务
// 统一管理白噪音和音乐播放，支持多音源叠加 + 交叉淡化

final class AudioService: ObservableObject {
    static let shared = AudioService()

    // ── 白噪音播放器（可叠加多个）──
    private var ambientPlayers: [String: AVAudioPlayer] = [:]

    // ── 音乐播放器 ──
    private var musicPlayer: AVAudioPlayer?
    @Published var isMusicPlaying: Bool = false

    // ── 音频会话 ──
    init() {
        configureAudioSession()
    }

    private func configureAudioSession() {
        let session = AVAudioSession.sharedInstance()
        do {
            try session.setCategory(.playback, mode: .default, options: [.mixWithOthers])
            try session.setActive(true)
        } catch {
            print("[AudioService] 音频会话配置失败: \(error)")
        }
    }

    // MARK: - 白噪音

    func playAmbient(_ fileName: String, volume: Float = 0.5) {
        guard ambientPlayers[fileName] == nil else { return }
        guard let url = Bundle.main.url(forResource: fileName, withExtension: "mp3") else {
            print("[AudioService] 未找到文件: \(fileName).mp3")
            return
        }

        do {
            let player = try AVAudioPlayer(contentsOf: url)
            player.numberOfLoops = -1  // 无限循环
            player.volume = 0
            player.play()
            ambientPlayers[fileName] = player

            // 渐入
            fadeIn(player, to: volume, duration: 2.0)
        } catch {
            print("[AudioService] 播放失败: \(error)")
        }
    }

    func stopAmbient(_ fileName: String) {
        guard let player = ambientPlayers[fileName] else { return }
        fadeOut(player, duration: 1.5) { [weak self] in
            player.stop()
            self?.ambientPlayers.removeValue(forKey: fileName)
        }
    }

    func setAmbientVolume(_ fileName: String, volume: Float) {
        ambientPlayers[fileName]?.volume = volume
    }

    func stopAllAmbient() {
        for (key, player) in ambientPlayers {
            fadeOut(player, duration: 1.0) { [weak self] in
                player.stop()
                self?.ambientPlayers.removeValue(forKey: key)
            }
        }
    }

    // MARK: - 音乐

    func playMusic(_ fileName: String) {
        musicPlayer?.stop()

        guard let url = Bundle.main.url(forResource: fileName, withExtension: "mp3") else {
            print("[AudioService] 未找到文件: \(fileName).mp3")
            return
        }

        do {
            musicPlayer = try AVAudioPlayer(contentsOf: url)
            musicPlayer?.volume = 0
            musicPlayer?.play()
            isMusicPlaying = true
            fadeIn(musicPlayer!, to: 1.0, duration: 1.5)
        } catch {
            print("[AudioService] 播放失败: \(error)")
        }
    }

    func toggleMusicPlayPause() {
        guard let player = musicPlayer else { return }
        if player.isPlaying {
            player.pause()
            isMusicPlaying = false
        } else {
            player.play()
            isMusicPlaying = true
        }
    }

    func stopMusic() {
        guard let player = musicPlayer else { return }
        fadeOut(player, duration: 1.0) { [weak self] in
            player.stop()
            self?.isMusicPlaying = false
        }
    }

    func seekMusic(to time: TimeInterval) {
        musicPlayer?.currentTime = time
    }

    // MARK: - 交叉淡化工具

    private func fadeIn(_ player: AVAudioPlayer, to targetVolume: Float, duration: TimeInterval) {
        let steps = 30
        let interval = duration / Double(steps)
        let increment = targetVolume / Float(steps)

        for i in 0...steps {
            DispatchQueue.main.asyncAfter(deadline: .now() + interval * Double(i)) {
                player.volume = min(increment * Float(i), targetVolume)
            }
        }
    }

    private func fadeOut(_ player: AVAudioPlayer, duration: TimeInterval, completion: @escaping () -> Void) {
        let steps = 20
        let interval = duration / Double(steps)
        let startVolume = player.volume
        let decrement = startVolume / Float(steps)

        for i in 0...steps {
            DispatchQueue.main.asyncAfter(deadline: .now() + interval * Double(i)) {
                player.volume = max(startVolume - decrement * Float(i), 0)
                if i == steps { completion() }
            }
        }
    }
}

File: satori_flutter/lib/services/audio_service.dart
import 'package:audioplayers/audioplayers.dart';
import 'media_session_service.dart';

/// 音频服务
/// 统一管理白噪音和音乐播放，支持多音源叠加 + 交叉淡化
class AudioService {
  AudioService._();
  static final instance = AudioService._();

  /// 可选的系统媒体会话服务（Phase P6）。
  /// 调用者在初始化时注入，听雨和抚琴后台播放时自动同步。
  MediaSessionService? mediaSession;

  // ── 白噪音播放器（可叠加多个）──
  final Map<String, AudioPlayer> _ambientPlayers = {};

  // ── 音乐播放器 ──
  AudioPlayer? _musicPlayer;
  bool isMusicPlaying = false;

  // MARK: - 白噪音

  Future<void> playAmbient(String fileName, {double volume = 0.5}) async {
    if (_ambientPlayers.containsKey(fileName)) return;

    final player = AudioPlayer();
    await player.setReleaseMode(ReleaseMode.loop);
    await player.setVolume(0);
    await player.play(AssetSource('audio/$fileName.mp3'));
    _ambientPlayers[fileName] = player;

    // 渐入
    _fadeIn(player, targetVolume: volume, duration: const Duration(seconds: 2));
  }

  Future<void> stopAmbient(String fileName) async {
    final player = _ambientPlayers[fileName];
    if (player == null) return;

    await _fadeOut(player, duration: const Duration(milliseconds: 1500));
    await player.stop();
    await player.dispose();
    _ambientPlayers.remove(fileName);
  }

  Future<void> setAmbientVolume(String fileName, double volume) async {
    await _ambientPlayers[fileName]?.setVolume(volume);
  }

  Future<void> stopAllAmbient() async {
    for (final entry in _ambientPlayers.entries.toList()) {
      await _fadeOut(entry.value, duration: const Duration(seconds: 1));
      await entry.value.stop();
      await entry.value.dispose();
    }
    _ambientPlayers.clear();
  }

  // MARK: - 音乐

  Future<void> playMusic(String fileName, {String? title, String? artist, Duration? duration}) async {
    await _musicPlayer?.stop();
    await _musicPlayer?.dispose();

    _musicPlayer = AudioPlayer();
    await _musicPlayer!.setVolume(0);
    await _musicPlayer!.play(AssetSource('audio/$fileName.mp3'));
    isMusicPlaying = true;
    _fadeIn(_musicPlayer!, targetVolume: 1.0, duration: const Duration(milliseconds: 1500));

    // 同步系统媒体会话
    mediaSession?.updateMetadata(
      title: title ?? fileName,
      artist: artist,
      duration: duration,
    );
    mediaSession?.updatePlaybackState(isPlaying: true);
  }

  Future<void> toggleMusicPlayPause() async {
    final player = _musicPlayer;
    if (player == null) return;

    if (isMusicPlaying) {
      await player.pause();
      isMusicPlaying = false;
    } else {
      await player.resume();
      isMusicPlaying = true;
    }
    mediaSession?.updatePlaybackState(isPlaying: isMusicPlaying);
  }

  Future<void> stopMusic() async {
    final player = _musicPlayer;
    if (player == null) return;

    await _fadeOut(player, duration: const Duration(seconds: 1));
    await player.stop();
    isMusicPlaying = false;
    mediaSession?.updatePlaybackState(isPlaying: false);
    mediaSession?.release();
  }

  Future<void> seekMusic(Duration position) async {
    await _musicPlayer?.seek(position);
  }

  Duration? get musicPosition => null; // Retrived via stream
  Stream<Duration>? get musicPositionStream => _musicPlayer?.onPositionChanged;
  Stream<void>? get musicCompleteStream => _musicPlayer?.onPlayerComplete;

  // MARK: - 交叉淡化

  Future<void> _fadeIn(
    AudioPlayer player, {
    required double targetVolume,
    required Duration duration,
  }) async {
    const steps = 30;
    final interval = duration.inMilliseconds ~/ steps;
    final increment = targetVolume / steps;

    for (var i = 1; i <= steps; i++) {
      await Future.delayed(Duration(milliseconds: interval));
      final vol = (increment * i).clamp(0.0, targetVolume);
      await player.setVolume(vol);
    }
  }

  Future<void> _fadeOut(
    AudioPlayer player, {
    required Duration duration,
  }) async {
    const steps = 20;
    final interval = duration.inMilliseconds ~/ steps;
    final startVolume = 1.0; // approximate
    final decrement = startVolume / steps;

    for (var i = 1; i <= steps; i++) {
      await Future.delayed(Duration(milliseconds: interval));
      final vol = (startVolume - decrement * i).clamp(0.0, 1.0);
      await player.setVolume(vol);
    }
  }
}
